Validation/HGCalValidation/src/HGVHistoProducerAlgo.cc

```diff
@@ -2504,94 +2497,99 @@ void HGVHistoProducerAlgo::multiClusters_to_CaloParticles(const Histograms& hist
+                                 << totalCPEnergyFromLayerCP << "\t" << std::setw(22) << energyFractionOfTSTinCP << "\t"
+                                 << std::setw(25) << energyFractionOfCPinTST << std::endl;
 
-    }  //end of loop through multi clusters
-  }
-  //Loop through multiclusters
-  for (unsigned int mclId = 0; mclId < nMultiClusters; ++mclId) {
-    const auto& hits_and_fractions = multiClusters[mclId].hitsAndFractions();
-    if (!hits_and_fractions.empty()) {
-      // find the unique caloparticles id contributing to the multi clusters
-      //cpsInMultiCluster[multicluster][CPids]
-      std::sort(cpsInMultiCluster[mclId].begin(), cpsInMultiCluster[mclId].end());
-      auto last = std::unique(cpsInMultiCluster[mclId].begin(), cpsInMultiCluster[mclId].end());
-      cpsInMultiCluster[mclId].erase(last, cpsInMultiCluster[mclId].end());
-
-      if (multiClusters[mclId].energy() == 0. && !cpsInMultiCluster[mclId].empty()) {
-        //Loop through all CaloParticles contributing to multicluster mclId.
-        for (auto& cpPair : cpsInMultiCluster[mclId]) {
-          //In case of a multi cluster with zero energy but related CaloParticles the score is set to 1.
-          cpPair.second = 1.;
-          LogDebug("HGCalValidator") << "multiCluster Id: \t" << mclId << "\t CP id: \t" << cpPair.first
-                                     << "\t score \t" << cpPair.second << std::endl;
-          histograms.h_score_multicl2caloparticle[count]->Fill(cpPair.second);
-        }
-        continue;
+    }  //end of loop through Trackster's layerClusters
+  }    //end of loop through Tracksters
+
+  //Loop through Tracksters
+  for (unsigned int tstId = 0; tstId < nTracksters; ++tstId) {
+    if (tracksters[tstId].vertices().empty())
+      continue;
+
+    // find the unique CaloParticles id contributing to the Tracksters
+    //cpsInTrackster[trackster][CPids]
+    std::sort(cpsInTrackster[tstId].begin(), cpsInTrackster[tstId].end());
+    auto last = std::unique(cpsInTrackster[tstId].begin(), cpsInTrackster[tstId].end());
+    cpsInTrackster[tstId].erase(last, cpsInTrackster[tstId].end());
+
+    if (tracksters[tstId].raw_energy() == 0. && !cpsInTrackster[tstId].empty()) {
+      //Loop through all CaloParticles contributing to Trackster tstId.
+      for (auto& cpPair : cpsInTrackster[tstId]) {
+        //In case of a Trackster with zero energy but related CaloParticles the score is set to 1.
+        cpPair.second = 1.;
+        LogDebug("HGCalValidator") << "Trackster Id: \t" << tstId << "\t CP id: \t" << cpPair.first << "\t score \t"
+                                   << cpPair.second << std::endl;
+        histograms.h_score_trackster2caloparticle[count]->Fill(cpPair.second);
       }
+      continue;
+    }
+
+    for (const auto lcId : tracksters[tstId].vertices()) {
+      const auto& hits_and_fractions = layerClusters[lcId].hitsAndFractions();
```

Comment:
I believe you are missing the correct normalization using the `vertex_multiplicity`, right?
LayerClusters are built **before** `Tracksters` and have no concept, so far, of energy sharing. This is added on top while building Tracksters and were propagated correctly while creating `MultiClustersFromTracksters`. We should do the same here.